File: LinkedList/Dart/linked_list.dart
class Node {
  Node(this.value);
  dynamic value;
  Node? next;
}

class LinkedList {
  LinkedList(dynamic value) {
    Node newNode = Node(value);
    this.head = newNode;
    this.tail = newNode;
    this.length = 1;
  }

  int length = 0;
  Node? head;
  Node? tail;

  //print the whole linked list
  bool? printList() {
    if (length == 0) {
      return null;
    }
    Node? temp = head!;
    while (temp != null) {
      print(temp.value);
      temp = temp.next;
    }
    return true;
  }

  //push new node to list
  void push(dynamic value) {
    if (length == 0) {
      LinkedList(value);
    } else {
      Node newNode = Node(value);
      tail!.next = newNode;
      tail = newNode;
      length++;
    }
  }

  //pop from end
  bool pop() {
    if (length == 0) {
      return false;
    } else {
      Node temp = head!;
      Node pre = head!;
      while (temp.next != null) {
        pre = temp;
        temp = temp.next!;
      }

      pre.next = null;
      tail = pre;
      length--;
      return true;
    }
  }

  //prepend listf
  void insertAtStart(dynamic value) {
    if (length == 0) {
      LinkedList(value);
    } else {
      Node newHead = Node(value);
      newHead.next = head;
      head = newHead;
      length++;
    }
  }

  bool popFirst() {
    if (length == 0) {
      return false;
    }
    Node? newHead = head!.next;
    head = newHead;
    length--;
    return true;
  }

  Node? getNode(int index) {
    if (length == 0 || index + 1 > length || index < 0) {
      return null;
    }
    // else if (index == 1) {
    //   return head;
    //}
    else {
      Node temp = head!;
      for (int i = 0; i < index; i++) {
        temp = temp.next!;
      }
      return temp;
    }
  }

  bool setIndex(dynamic value, int index) {
    Node? node = getNode(index);
    if (node != null) {
      node.value = value;
      return true;
    } else {
      return false;
    }
  }

  bool insertAtIndex(dynamic value, int index) {
    if (index < 0 || index > length) {
      return false;
    }
    if (index == 0) {
      insertAtStart(value);
    } else if (index == length) {
      push(value);
    } else {
      Node node = Node(value);
      Node temp = head!;
      for (var i = 0; i < index - 1; i++) {
        temp = temp.next!;
      }
      node.next = temp.next;
      temp.next = node;
    }
    return true;
  }

  bool removeItem(int index) {
    if (index < 0 || index >= length) {
      return false;
    }
    if (index == 0) {
      popFirst();
    } else if (index == length - 1) {
      pop();
    } else {
      Node temp = head!;
      for (int i = 0; i < index - 1; i++) {
        temp = temp.next!;
      }
      Node removed = temp.next!;
      Node? next = removed.next;
      temp.next = next;
    }
    return true;
  }
}

void main() {
  LinkedList myLinkedList = LinkedList(4);
  myLinkedList.push(5);
  myLinkedList.push(10);
  myLinkedList.push(20);
  myLinkedList.removeItem(3);
  myLinkedList.printList();
}

File: LinkedList/Python/linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self, value):
        newNode = Node(value)
        self.head = newNode
        self.tail = newNode
        self.length = 1

    def printList(self):
        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next

    def append(self, value):
        if self.head is None:
            self.__init__(value)
        else:
            newNode = Node(value)
            self.tail.next = newNode
            self.tail = newNode
        self.length += 1
        return True

    def pop(self):
        if self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            temp = self.head
            pre = self.head
            while temp.next is not None:
                pre = temp
                temp = temp.next
            pre.next = None
            self.tail = pre
        self.length -= 1
        return True

    def prepend(self, value):
        if self.length == 0:
            self.__init__(value)
        else:
            newNode = Node(value)
            oldHead = self.head
            self.head = newNode
            self.head.next = oldHead
        self.length += 1
        return True

    def popFirst(self):
        if self.length == 0:
            return None
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
        self.length -= 1
        return True

    def get(self, index: int):
        if self.length == 0 or self.length <= index or index < 0:
            return None
        else:
            val = self.head
            for i in range(index):
                val = val.next
            return val

    def set_value(self, index: int, value):
        node = self.get(index)
        if node:
            node.value = value
            return True
        else:
            return False

    def insert(self, index: int, value):
        if index < 0 or index > self.length:
            return False
        elif index == 0:
            return self.prepend(value)
        elif index == self.length:
            return self.append(value)
        else:
            newNode = Node(value)
            temp = self.get(index - 1)
            newNode.next = temp.next
            temp.next = newNode
        return True

    def remove_item(self, index: int):
        if index < 0 or index + 1 > self.length:
            return False
        elif index == 0:
            return self.popFirst()
        elif index == self.length - 1:
            return self.pop()
        else:
            prevItem = self.get(index - 1)
            item = prevItem.next
            prevItem.next = item.next
        return True


myLinkedList = LinkedList(4)

myLinkedList.append(10)
myLinkedList.append(45)

myLinkedList.printList()

myLinkedList.remove_item(1)
print('after')

myLinkedList.printList()



